packages/sanity/src/core/studio/copyPaste/useCopyPasteAction.ts

```diff
@@ -0,0 +1,333 @@
+
+      const schemaTypeAtPath = resolveSchemaTypeForPath(schemaType, path, value)
+
+      if (!schemaTypeAtPath) {
+        toast.push({
+          status: 'error',
+          title: t('copy-paste.on-copy.validation.schema-type-incompatible.title', {
+            path: PathUtils.toString(path),
+          }),
+        })
+        return
+      }
+
+      const isDocument = schemaTypeAtPath.type?.name === 'document'
+      const isArray = schemaTypeAtPath.jsonType === 'array'
+      const isObject = schemaTypeAtPath.jsonType === 'object'
+      const valueAtPath = getValueAtPath(value, path)
+
+      // Test if the value is empty (undefined, empty object or empty array)
+      if (isEmptyValue(valueAtPath)) {
+        toast.push({
+          status: 'warning',
+          title: t('copy-paste.on-copy.validation.no-value.title'),
+        })
+        return
+      }
+
+      const payloadValue: CopyActionResult = {
+        _type: 'copyResult',
+        documentId,
+        documentType,
+        isDocument,
+        isArray,
+        isObject,
+        items: [
+          {
+            schemaTypeName: schemaTypeAtPath.name || 'unknown',
+            schemaTypeTitle: schemaTypeAtPath.title || schemaType.name || 'unknown',
+            documentPath: path,
+            value: valueAtPath,
+          },
+        ],
+      }
+
+      telemetry.log(FieldCopied, {
+        context: options?.context?.source || 'unknown',
+        schemaTypes: [schemaTypeAtPath.jsonType],
+      })
+
+      setCopyResult(payloadValue)
+      const isWrittenToClipboard = await writeClipboardItem(payloadValue)
+
+      if (!isWrittenToClipboard) {
+        toast.push({
+          status: 'error',
+          title: t('copy-paste.on-copy.validation.clipboard-not-supported.title'),
+        })
+        return
+      }
+
+      const fields = payloadValue.items.map((item) => item.schemaTypeTitle).join('", "')
+
+      toast.push({
+        status: 'success',
+        title: isDocument
+          ? t('copy-paste.on-copy.validation.copy-document-success.title', {
+              fieldNames: fields,
+            })
+          : t('copy-paste.on-copy.validation.copy-field_one-success.title', {
+              fieldName: fields,
+            }),
+      })
+    },
+    [documentMeta, setCopyResult, telemetry, toast, t],
+  )
+
+  const onPaste = useCallback(
+    async (targetPath: Path, value: FormDocumentValue | undefined, options?: PasteOptions) => {
+      const {schemaType: targetDocumentSchemaType} = documentMeta || {}
+      const targetSchemaType = resolveSchemaTypeForPath(
+        targetDocumentSchemaType!,
+        targetPath,
+        value,
+      )!
+
+      const clipboardItem = await getClipboardItem()
+
+      // Return early if no clipboard item or if clipboard item is invalid
+      if (!clipboardItem) {
+        toast.push({
+          status: 'info',
+          title: t('copy-paste.on-paste.validation.clipboard-empty.title'),
+        })
+        return
+      }
```

Comment:
similar comment to the one i left below, i think i would prefer the `paste field` button to be disabled or just not appear if there is nothing to paste instead of seeing this toast

Reply:
We discussed this but landed on not wanting these be disabled, but rather give the user feedback.
In general i'd say its bad UX wise to disable things without giving enough context to the user. We also didn't find a performant way to disable field actions. Last point, you can't actually read the clipboard outside a user action to check if you have something in clipboard, so don't think there is good options there at all.